### PR TITLE
Lint JavaScript class methods and properties

### PR DESCRIPTION
**Pull request by wbamberg**
This PR will eventually be to add linting for all the JS class methods and properties, to fix https://github.com/mdn/stumptown-content/issues/348.

At the moment it only lints `data.static_methods` but I'd love some feedback at this point anyway :).

************

The [first commit](https://github.com/mdn/stumptown-content/commit/02df9ba074d4d0019597e52539fa83cc0ec41e58) is a relatively minimal thing, that really does three things:
* updates the main index.js to lint for all ingredients, not just mandatory ones
* updates `logMissingIngredient` to `logIngredientError`, so we can use it to log errors about optional ingredients (because missing optional ingredients are allowed)
* add the handler for `data.static_methods?`. We could argue about exactly which checks it makes, but this seemed like some useful stuff.

This will usually log errors for the current docs, because it wants to see `<a>` elements in the `<dt>` elements, and most of the current docs use `{{jsxref}}`. But since we are aiming to remove these, that ought to be OK. I de-jsxref'd [`BigInt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) and [`Array`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array), so you can try it on them. And I've also been using https://wiki.developer.mozilla.org/en-US/docs/User:wbamberg/test-linter2 to test it out.

I also factored some error logging code into its own [`recordError`](https://github.com/mdn/stumptown-content/commit/02df9ba074d4d0019597e52539fa83cc0ec41e58#diff-0e8d23035d4a145bd6a3eedd19e0ab00R12) function, as this seemed very repetitive. I think this is an incomplete thought in that commit.

***********

But, I think linting things that are a `<dl>` of links are going to be a common thing in various ingredient linters, so I wanted to factor it out, and that's what the [second commit](https://github.com/mdn/stumptown-content/commit/792c960c0e60fa7c61f503a20fd8ff6ed8172b10) does. But at that point I also got tired of having to schlep all this state around for reporting errors, so I built `recordError` into utils.js, and renamed it [`logError`](https://github.com/mdn/stumptown-content/commit/792c960c0e60fa7c61f503a20fd8ff6ed8172b10#diff-5b94f8a37b1f54bfd803a368d3654b17R74). If we like this approach we could use it in places like [this](https://github.com/mdn/stumptown-content/blob/master/scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-browser-compatibility.js#L15-L21) and [this](https://github.com/mdn/stumptown-content/blob/master/scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-specifications.js#L15-L21).

Anyway. I'm not sure if the things I've done to the error reporting are good. But I feel as if at the moment error reporting is quite repetitive and involves a lot of state, so it might be worth thinking about something like this.

### CODE REVIEW

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-handler.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+          "Links in link lists must be listed in alphabetical order",
+          "link-list-alpha-order"
+        );
+      }
+      title = code.children[0].value;
+    }
+  }
+}
+
+function checkLinkList(section, logError) {
```
Comment (ddbeck) [inline scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-handler.js]: I'm finding this function a little hard to follow. I guess we're trying to figure out if the section is well-formed, but this code is a bit shy about saying what that is (we check text nodes and elements, but for what? Nothing really says). I've inferred that a section should:

* Contain one `<h2>` element (which we've already cut off)
* Contain one or more `<dl>` elements, such that each `<dl>` element:
  * Contains zero or more `<dt>` elements in alphabetical order (based on their text content)
  * Each `<dt>`'s children are of the form `<a><code>textLiteral</code></a>`
* Contain only white space text literals
* Contain no other elements

Do I have that right? I think it might help if this were structured to show these requirements a bit more directly. It might be slightly more expensive, but, you could, for example:

1. Check section.children for forbidden element types and text literals
2. Visit each `<dl>`, collect the text of each `<dt>`, and check the alpha sort
3. Visit each `<dt>` and check its well-formedness.

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-handler.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+/**
+ * Check the structure of a `dl` element that represents a list of links.
+ */
+function checkDL(node, logError) {
+  let title = "";
+  for (const child of node.children) {
+    // We are only going to check `dt` elements, and will tolerate
+    // any others
```
Comment (ddbeck) [inline scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-handler.js]: There are some utilities that might make traversing these trees a bit easier:

Using [`unist-util-visit`](https://github.com/syntax-tree/unist-util-visit#readme) or `hast-util-select`'s `selectAll` and `visit` might be a more idiomatic way to walk all these elements. Something like…

```javascript
visit(tree, node => node.tagName === "dt", node => /* … do something with each dt … */);
```

or

```javascript
const dts = selectAll('dt', tree);
for (const dt of dts) {
  visit(dt.children, node => {
    /* .. */
  });
}
```

[`unist-util-visit-parents`](https://github.com/syntax-tree/unist-util-visit-parents) will let you visit the innermost element you're looking for and then check bottom-up for the structure you're looking for:

```javascript
visit(dtNode, node => node.type === "text", (node, ancestors) => {
  /* assert !children.length and the immediate ancestors are `<code>` and `<a>` */
});
```

(There might be a util to get the leaf node(s) of a tree too, don't know off the top of my head.)

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  // This is an optional ingredient, so if there's no `h2`,
+  // assume that the page intends to omit it.
+  if (heading === null) {
+    return;
+  }
+
+  const section = utils.sliceSection(heading, body);
+  // the first element is always the `h2`, which we are not interested in
+  const children = section.children.slice(1);
+  const logError = utils.logError(body, file, context);
```
[the remark this answers is not in the document]
Comment (wbamberg) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js]: > there's a more concrete problem

Oh, OK. I hadn't realised that this was what this argument was for. I guess that's evidence that error reporting here is quite complex. Of course this would be easy to fix by making the node an argument to the logging function (like `text` and `id` are now) rather than an argument to the factory: that is, it's a thing you know at the point you log the error.

But on the general points: fair enough if this seems too complex. But I think there is a problem in this code at the moment. First, as I understand it, lots of places are going to have to write code like:

```js
const message = file.message(
  `Expected h2#${id}`,
  body,
  `${context.source}:${context.recipeName}/${context.ingredient}/expected-heading`
);
message.fatal = true;
utils.logMissingIngredient(file, context);
```

So that's a lot of duplication, and a lot of scope for error messages to be inconsistently formatted from one place to another.

Second, unless I misunderstand this, anywhere that's doing this is going to need the `file` and `context` objects. So at the moment we're passing these things around. And if we want to factor common bits of linting into separate modules, we have to pass them around there as well. It's just a lot of state to ship around.

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-handler.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+function checkLinkList(section, logError) {
+  for (const node of section) {
+    switch (node.type) {
+      case "text":
+        checkTextNode(node, logError);
+        break;
+      case "element":
+        checkElementNode(node, logError);
+        break;
+      default:
```
[the remark this answers is not in the document]
Comment (wbamberg) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-handler.js]: I guess it could in theory be any of these: https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#Node_type_constants. It seems like it might be good to know about some of them.

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js]: Yes, I agree that the places where we call `file.message()` are high risk for inconsistency and that we're currently passing around a lot of state. You might be on the right track of creating a logger at the top of a handler and passing it around, but maybe we can work on the API a bit. Something like this?

```javascript
const logger = new IngredientLogger(file, context);
logger.fail() // fails the whole ingredient, using the whole file as a position, like utils.logMissingIngredient
logger.fail(someNode) // fails the whole ingredient, using a specific node as a position, such as an h2
logger.expected(someNode, "compat-macro") // logs an "expected-compat-macro" message against someNode, then calls `logger.fail` too
```

Or a more half-baked idea I've had would be to make all the logging happen at the top of the plugin and have ingredient handlers throw an exception type. Then you could write some module-level `Error` that the plugin would turn into a vfile message.

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-handler.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-handler.js]: Ah, OK. This is a confusing bit about unified. These are [hast nodes](https://github.com/syntax-tree/hast#nodes), not regular DOM nodes. I think the way the tree has been sliced already means that these nodes probably won't ever be anything but `text`, `element`, or `comment` (unless someone decides to put a doctype declaration in the middle of a page).

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (wbamberg) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js]: How do you want to proceed with this?

1. refactor logging as part of this PR  
2. separate PRs:
    * first, continue with this PR using the current logging
    * then refactor it in a follow-up
3. separate PRs:
    * first, refactor logging
    * then continue with this PR

The problem with (1): bigger PR, less focused, more disparate discussions.
The problem with (2): this PR will get ugly without some attempt to tame the logging code.
The problem with (3): this PR is pushed further back.

I think (3) is probably best.

[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js]: Yeah, let's refactor logging, then go from there. I'll plan to review as quickly as possible to minimize the delay.

[the remark this answers is not in the document]
Comment (wbamberg) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js]: I've filed https://github.com/mdn/stumptown-content/issues/385 to track the logging issues.

[the remark this answers is not in the document]
Comment (wbamberg) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js:23]: Resolved via https://github.com/mdn/stumptown-content/pull/386

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+const { select } = require("hast-util-select");
+
+const utils = require("./utils.js");
+const checkLinkList = require("./link-list-handler.js");
+
+/**
+ * Handler for the `data.specifications` ingredient.
```
[the remark this answers is not in the document]
Comment (wbamberg) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js]: -> https://github.com/mdn/stumptown-content/pull/374/commits/b5843801151e5e1777310ee219a447c00c0b4c4d

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+ */
+function handleDataStaticMethods(tree, file, context) {
+  const id = "Static_methods";
+  const body = select(`body`, tree);
```
[the remark this answers is not in the document]
Comment (wbamberg) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-static-methods.js]: -> https://github.com/mdn/stumptown-content/pull/374/commits/ca13bd475b2e89dfc47d50b23566f5c3464d178c

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-handler.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
+/**
```
[the remark this answers is not in the document]
Comment (wbamberg) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-handler.js]: -> https://github.com/mdn/stumptown-content/pull/374/commits/5b97c3d5743f9e0b0fb3f846a18c0cc6f9f9bcbc

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-handler.js (hunk 1 above)
Comment (wbamberg) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-handler.js]: I've substantially rewritten this to be I hope more along the lines you have suggested here. I didn't use `visit` because to be honest I didn't think it would make this code any clearer.

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-checker.js [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        dtCode,
+        "code elements in dt elements in link lists must contain a single text node",
+        "only-single-text-node-element-in-link-list-code"
+      );
+    }
+  }
+
+  // <dt><dd> pairs must be ordered by the alphabetical order of
+  // the <dt><a><code> text content
+  let title = "";
```
Comment (ddbeck) [inline scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-checker.js:87]: Maybe this could be `prevTitle`? I had to re-read this a couple times to figure out how this fit into the loop.

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-class-members.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+/**
+ * Handler for the `data.static_methods` ingredient.
+ */
+function handleDataStaticMethods(tree, logger) {
+  checkLinkList("Static_methods", tree, logger);
+}
+
+module.exports = {
+  handleDataStaticMethods,
```
[the remark this answers is not in the document]
Comment (wbamberg) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-class-members.js:11]: Oh, yes, I meant to finish this bit off. The idea would be to have a separate `handleXYZ` for each class member ingredient (e.g. `data.static_properties`, `data.instance_methods` etc) that just calls the underlying function, passing in the different ID. Then we export all of those from this module. Does that sound good? If so I can add it now.

[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-class-members.js:11]: Oh yeah, that does sound nice. 👍

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-checker.js (hunk 8 above)
Comment (wbamberg) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/link-list-checker.js]: -> https://github.com/mdn/stumptown-content/pull/374/commits/cacf81c5e3ce1c16055b94c5c0ebaf2900576c99

(apparently I also forgot to change the name in my `require` when I renamed the utility module)

File: scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-class-members.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (ddbeck) [reply scripts/scraper-ng/rules/html-require-recipe-ingredients/ingredient-handlers/data-class-members.js:40]: This is fine. Thank you! 👍
